loopy/codegen/__init__.py

```diff
@@ -550,4 +550,23 @@ def generate_body(kernel):
 
 # }}}
 
-# vim: foldmethod=marker
+from loopy.target.c import CASTIdentityMapper
+class FunctionDeclExtractor(CASTIdentityMapper):
+    def __init__(self):
+        self.decls = []
+
+    def map_function_declaration(self, node):
+        self.decls.append(node)
+        return super(self.__class__, self).map_function_declaration(node)
+
+def generate_header(kernel):
```

Comment:
I think my solution of moving the extractor to CTarget, and pairing with a NonImplementedError in the TargetBase / warning if an invalid target is used in the generate_header function solves this... however, I don't know that `get_device_decl_extractor` function in the Target's make a ton of sense.  Is there a better place to have this?